### ISSUE TITLE
udev-builtin-blkid always uses BLKID_SUBLKS_BADCSUM flag

### ISSUE THREAD

**Issue by n3world** [systemd/systemd#6110]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 231

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

> Fedora

### Expected behaviour you didn't see

> builtin_blkid is able to detect the ext4 file system and attributes

### Unexpected behaviour you saw

> builtin_blkid was not able to detect the ext4 filesystem and the blkid function blkid_do_safeprobe returned -2, ambivalent result, when called by builtin_blkid. The command blkid -p <device> was able to correctly detect the ext4 file system.

### Steps to reproduce the problem

> I do not know how to reproduce the issue.

The only real difference between blkid -p and builtin_blkid is the use of the BLKID_SUBLKS_BADCSUM, which was added in commit d47f6ca5f9b7a0b400d8bdb050151a0284fb4bdb. My guess of what happened is libblkid detected at least two superblocks on the device only one of which had a good checksum and the others had a bad checksum. When the flag BLKID_SUBLKS_BADCSUM was used blkid returns -2 but without that flag it can correctly detect the ext4 filesystem.

One simple fix is to just remove the flag and the handling of bad checksum but if it is really desired it could be kept as a second attempt if the first probe returns nothing found.

**Comment by poettering** [comment]
Hmm, not sure I follow, but does your ext4 partition have an incorrect superblock checksum? We are usually very conservative with detecting file systems and if there's indication of something not being a valid superblock we shy away of tagging it as one.

Isn't the better approach that you fix your superblock checksum to be valid?

**Comment by n3world** [comment]
No the ext4 partition was fine. blkid thought there was  another partition there which I am guessing had a bad checksum since without the flag BLKID_SUBLKS_BADCSUM blkid only recognized the ext4 partition. fsck passed on the ext4 partition and I was able to mount it fine.

Unfortunately I do not know what else blkid detected on the partition. I have had issue in the past were blkid detected two different types. In that case it was swap and minix. In that case the minix magic number happens to be the same location as the swap UUID so depending on the swap UUID blkid would get confused and couldn't tell if it was minix or swap. In this case the filesystem structure or something else must have just lined up with another partition type but with a bad checksum so that blkid when it ignored checksums did not know what was there.

When I ran the blkid command line, which does not use the flag BLKID_SUBLKS_BADCSUM, it could correctly tell that the partition was  ext4.

**Comment by poettering** [comment]
well, if there's ambiguity we don't tag on purpose, since automatically making use of a partition that might actually be something else is possibly destructive, and if we have a hint that something might be wrong, we shouldn't do anything.

I am pretty sure the current behaviour is actually correct: don't do anything automatically on partitions that are ambiguous.

Such ambigities happen, primarily due to the poor design of the on-disk formats of some file systems. To deal with this util-linux contains an explicit tool even: wipefs, which can properly remove detection ambiguities like yours. https://karelzak.blogspot.de/2009/11/wipefs8.html

I am going to close this bug now, as I think the correct way out of this is to use a tool such as wipefs and remove the ambiguity rather than tape over it, possibly risking destructive automatic behaviour on other instances like this.

Sorry if that's disappointing, but I hope this makes some sense!

**Comment by n3world** [comment]
I don't think you understand what I was trying to say. If the flag BLKID_SUBLKS_BADCSUM is **not** used there is no ambiguity in blkid. The ambiguity is only introduced by using the flag BLKID_SUBLKS_BADCSUM.

This was a system which had never had anything else installed on it and after a clean reboot builtin_blkid could not detect the partition type but blkid the command line could.

Since the blkid command line worked fine it seems like it is more an issue with how builtin_blkid is using libblkid.

Why do you think the BLKID_SUBLKS_BADCSUM flag should be used on all superblock probes?

**Comment by poettering** [comment]
Ah, now I get it. So we should drop BLKID_SUBLKS_BADCSUM altogether.

**Comment by n3world** [comment]
Yes that is one possibility. The other being first try the probe without BLKID_SUBLKS_BADCSUM and if that finds nothing try it with BLKID_SUBLKS_BADCSUM and keep the warnings about bad checksums. I wasn't sure if the warnings were considered useful and wanted to be kept.

I can submit a PR that does either. I wanted to know which was preferred. Removing it is obviously the easiest thing.

Sorry my early explanations did not clearly convey the issue.

**Comment by poettering** [comment]
Hmm, actually thinking about this, this might actually be something libblkid should fix, not us. i.e.: if the checksum check is off, and libblkid detects one label with a good checksum and one with a bad, then it should not consider that an ambiguity but simply ignore the one with the bad checksum. However, if there are two labels detected, either both with a good or both with a bad checksum, then it should consider it an ambiguity.

@karelzak what do you say, isn't this something libblkid should fix, rather than us?

**Comment by karelzak** [comment]
I'll think about it. It is not so big problem to add another rule to the "safe probe" libblkid probing method.

Anyway, I'm not sure if enabled BLKID_SUBLKS_BADCSUM for udev is a good idea. Maybe this flags should be completely ignored by libblkid if "safe mode" (blkid_do_safeprobe()), because it's not safe to assume that superblock with bad checksum is valid. The flag has been introduced for tools like wipefs where we want maximal promiscuity.

The udev uses BLKID_SUBLKS_BADCSUM to report/log bad checksums, but for libblkid it also means "accept invalid superblock". 

Maybe the best solution would be to extend libblkid  API and add a way how to inform applications about bad checksums (or another issues) without impact to the probing result. Something like:
```
/* do probing as usually */
...

/* report issues */
while (blkid_probe_next_log(pr, &log) == 0) {
    if (blkid_log_has_value(log, "SBBADCSUM"))
            log_warning("incorrect %s checksum on %s",
                      blkid_log_get_value(log, "TYPE")
                      udev_device_get_devnode(dev));
}
```
As temporary solution I suggest to remove BLKID_SUBLKS_BADCSUM flag from udev.

**Comment by poettering** [comment]
@karelzak we don't actually make use of the probing data when the checksum is bad, it's just about printing a friendly log message, see this patch:

https://github.com/systemd/systemd/commit/d47f6ca5f9b7a0b400d8bdb050151a0284fb4bdb

And while that isn't crucial I think it's actually kind of a nice feature to have... except that it doesn't work this way right now, since it creates ambiguities where there shouldnt be amiguities...

Anyway, we can certainly also drop the logging from udev, but I think it would be much much nicer if blkid would fix that for us, so that we can keep the logging in.

Anyway, please let me know how you decide. If you don't want to do that in libblkid I'll drop these lines from udev.

**Comment by karelzak** [comment]
I know that udev does not use it, but libblkid uses the result to evaluate possible collisions, interactions with partition tables, etc. That's bad (for udev).

Please, remove the flag from the udev for now. 

I agree that the udev logging feature is a nice thing, but libblkid has to provide better way how to do it. We need something that will not affect probing logic within libblkid. If libblkid will provide a better way than we can report more things that only bad checksums (obsolete superblock, etc.)

Thanks.

**Comment by poettering** [comment]
Fix waiting in #6215
